File: src/services/waitingListService.js
import { api } from './api';

/**
 * Bekleme listesi ile ilgili işlemleri yöneten servis
 */
const waitingListService = {
  /**
   * Bir etkinliğin bekleme listesini getirir
   * @param {number} eventId - Etkinlik ID
   * @returns {Promise} - Bekleme listesi
   */
  getWaitingListByEvent: (eventId) => {
    return api.event.getWaitingList(eventId)
      .then(response => response.data);
  },
  
  /**
   * Bekleme listesindeki kullanıcıları getirir
   * @param {number} eventId - Etkinlik ID
   * @returns {Promise} - Kullanıcı listesi
   */
  getUsersInWaitingList: (eventId) => {
    return api.event.getWaitingList(eventId)
      .then(response => response.data);
  },
  
  /**
   * Kullanıcıyı bekleme listesine ekler
   * @param {number} eventId - Etkinlik ID
   * @param {number} userId - Kullanıcı ID
   * @param {string} note - İsteğe bağlı not
   * @returns {Promise} - Bekleme listesi kaydı
   */
  addUserToWaitingList: (eventId, userId, note = '') => {
    return api.event.addToWaitingList(eventId, { userId, note })
      .then(response => response.data);
  },
  
  /**
   * Kullanıcıyı bekleme listesinden çıkarır
   * @param {number} eventId - Etkinlik ID
   * @param {number} userId - Kullanıcı ID
   * @returns {Promise} - İşlem sonucu
   */
  removeUserFromWaitingList: (eventId, userId) => {
    return api.event.removeFromWaitingList(eventId, { userId })
      .then(response => response.data);
  },
  
  /**
   * Bekleme listesinin sırasını değiştirir
   * @param {number} eventId - Etkinlik ID
   * @param {Array} waitingList - Yeni sıralama ile bekleme listesi
   * @returns {Promise} - Güncellenen bekleme listesi
   */
  reorderWaitingList: (eventId, waitingList) => {
    return api.event.reorderWaitingList(eventId, waitingList)
      .then(response => response.data);
  },
  
  /**
   * Kullanıcının bütün bekleme listelerini getirir
   * @param {number} userId - Kullanıcı ID
   * @returns {Promise} - Kullanıcının bekleme listeleri
   */
  getUserWaitingLists: (userId) => {
    return api.event.getUserWaitingLists(userId)
      .then(response => response.data)
      .catch(error => {
        // 404 hatası kullanıcının hiç bekleme listesi olmadığını gösterir
        if (error.response && error.response.status === 404) {
          return [];
        }
        throw error;
      });
  },
  
  /**
   * Kullanıcının bekleme listesi durumunu kontrol eder
   * @param {number} eventId - Etkinlik ID
   * @param {number} userId - Kullanıcı ID
   * @returns {Promise} - Bekleme listesi durumu
   */
  checkUserWaitingStatus: (eventId, userId) => {
    return api.event.checkUserWaitingStatus(eventId, userId)
      .then(response => response.data);
  },
  
  /**
   * Bekleme listesinden katılımcıya yükseltir
   * @param {number} eventId - Etkinlik ID
   * @param {number} userId - Kullanıcı ID
   * @returns {Promise} - İşlem sonucu
   */
  promoteToParticipant: (eventId, userId) => {
    return api.event.promoteToParticipant(eventId, userId)
      .then(response => response.data);
  },
  
  /**
   * Otomatik olarak bekleme listesindeki kullanıcıları katılımcıya yükseltir
   * @param {number} eventId - Etkinlik ID
   * @param {number} limit - Maksimum yükseltilecek kişi sayısı
   * @returns {Promise} - Yükseltilen kişi sayısı
   */
  autoPromoteFromWaitingList: (eventId, limit = 0) => {
    return api.event.autoPromoteFromWaitingList(eventId, limit)
      .then(response => response.data);
  },
  
  /**
   * Etkinlik detaylarını getirir
   * @param {number} eventId - Etkinlik ID
   * @returns {Promise} - Etkinlik detayları
   */
  getEventDetails: (eventId) => {
    return api.event.getEventById(eventId)
      .then(response => response.data);
  },
  
  /**
   * Kullanıcıyı bekleme listesinden onaylayarak etkinliğe ekler
   * @param {number} eventId - Etkinlik ID
   * @param {number} userId - Kullanıcı ID
   * @returns {Promise} - İşlem sonucu
   */
  approveUserFromWaitingList: (eventId, userId) => {
    return api.event.approveUserFromWaitingList(eventId, userId)
      .then(response => response.data);
  }
};

export const {
  getWaitingListByEvent,
  getUsersInWaitingList,
  addUserToWaitingList,
  removeUserFromWaitingList,
  reorderWaitingList,
  getUserWaitingLists,
  getEventDetails,
  approveUserFromWaitingList
} = waitingListService;

export default waitingListService; 

File: cmManagementSystem/src/main/java/com/example/cmManagementSystem/service/WaitingListService.java
package com.example.cmManagementSystem.service;

import com.example.cmManagementSystem.dto.UserDto;
import com.example.cmManagementSystem.dto.WaitingListDto;
import com.example.cmManagementSystem.entity.WaitingList;

import java.time.LocalDateTime;
import java.util.List;

public interface WaitingListService {
    
    /**
     * Tüm bekleme listesi kayıtlarını getirir
     * 
     * @return Bekleme listesi kayıtları
     */
    List<WaitingListDto> findAll();
    
    /**
     * ID'ye göre bekleme listesi kaydı getirir
     * 
     * @param id Bekleme listesi kaydı ID
     * @return Bekleme listesi kaydı
     */
    WaitingListDto findById(Long id);
    
    /**
     * Etkinliğe göre bekleme listesi kayıtlarını getirir
     * 
     * @param eventId Etkinlik ID
     * @return Bekleme listesi kayıtları
     */
    List<WaitingListDto> findByEventId(Long eventId);
    
    /**
     * Etkinliğe göre bekleme listesi kayıtlarını getirir
     * WaitingListController için kullanılacak
     * 
     * @param eventId Etkinlik ID
     * @return Bekleme listesi kayıtları
     */
    List<WaitingListDto> findByEvent(Long eventId);
    
    /**
     * Etkinlik bekleme listesindeki kullanıcıları getirir
     * 
     * @param eventId Etkinlik ID
     * @return Kullanıcı listesi
     */
    List<UserDto> findUsersByEvent(Long eventId);
    
    /**
     * Etkinliğe ve duruma göre bekleme listesi kayıtlarını getirir
     * 
     * @param eventId Etkinlik ID
     * @param status Bekleme listesi durumu
     * @return Bekleme listesi kayıtları
     */
    List<WaitingListDto> findByEventIdAndStatus(Long eventId, WaitingList.WaitingStatus status);
    
    /**
     * Kullanıcıya göre bekleme listesi kayıtlarını getirir
     * 
     * @param userId Kullanıcı ID
     * @return Bekleme listesi kayıtları
     */
    List<WaitingListDto> findByUserId(Long userId);
    
    /**
     * Etkinlik ve kullanıcıya göre bekleme listesi kaydı getirir
     * 
     * @param eventId Etkinlik ID
     * @param userId Kullanıcı ID
     * @return Bekleme listesi kaydı
     */
    WaitingListDto findByEventIdAndUserId(Long eventId, Long userId);
    
    /**
     * Etkinlik ve kullanıcıya göre bekleme listesi kaydı getirir
     * WaitingListController için kullanılacak
     * 
     * @param eventId Etkinlik ID
     * @param userId Kullanıcı ID
     * @return Bekleme listesi kaydı
     */
    WaitingListDto findByEventAndUser(Long eventId, Long userId);
    
    /**
     * Etkinliğe göre bekleme listesi kayıtlarını sıra numarasına göre sıralı getirir
     * 
     * @param eventId Etkinlik ID
     * @return Bekleme listesi kayıtları
     */
    List<WaitingListDto> findByEventIdOrderByPositionAsc(Long eventId);
    
    /**
     * Etkinlik için beklemedekileri sayar
     * 
     * @param eventId Etkinlik ID
     * @param status Bekleme durumu
     * @return Bekleme listesindeki kişi sayısı
     */
    Integer countByEventIdAndStatus(Long eventId, WaitingList.WaitingStatus status);
    
    /**
     * Yeni bekleme listesi kaydı oluşturur
     * 
     * @param waitingListDto Bekleme listesi bilgileri
     * @return Oluşturulan bekleme listesi kaydı
     */
    WaitingListDto createWaitingList(WaitingListDto waitingListDto);
    
    /**
     * Kullanıcıyı bekleme listesine ekler
     * 
     * @param eventId Etkinlik ID
     * @param userId Kullanıcı ID
     * @param note Kullanıcı notu (isteğe bağlı)
     * @return Oluşturulan bekleme listesi kaydı
     */
    WaitingListDto addToWaitingList(Long eventId, Long userId, String note);
    
    /**
     * Bekleme listesi bilgilerini günceller
     * 
     * @param id Bekleme listesi ID
     * @param waitingListDto Güncellenecek bilgiler
     * @return Güncellenen bekleme listesi kaydı
     */
    WaitingListDto updateWaitingList(Long id, WaitingListDto waitingListDto);
    
    /**
     * Bekleme listesi durumunu günceller
     * 
     * @param id Bekleme listesi ID
     * @param status Yeni durum
     * @return Güncellenen bekleme listesi kaydı
     */
    WaitingListDto updateStatus(Long id, WaitingList.WaitingStatus status);
    
    /**
     * Bekleme listesi bildirim gönderildi olarak işaretler
     * 
     * @param id Bekleme listesi ID
     * @param notificationDate Bildirim tarihi
     * @param responseDeadline Yanıt tarihi
     * @return Güncellenen bekleme listesi kaydı
     */
    WaitingListDto markAsNotified(Long id, LocalDateTime notificationDate, LocalDateTime responseDeadline);
    
    /**
     * Bildirim gönderilmemiş ve bekleme durumundaki kayıtları getirir
     * 
     * @return Bekleme listesi kayıtları
     */
    List<WaitingListDto> findPendingNotifications();
    
    /**
     * Yanıt süresi dolmuş bildirim gönderilmiş kayıtları getirir
     * 
     * @param now Şu anki zaman
     * @return Süresi dolmuş bekleme listesi kayıtları
     */
    List<WaitingListDto> findExpiredNotifications(LocalDateTime now);
    
    /**
     * Bekleme listesi kaydını siler
     * 
     * @param id Silinecek bekleme listesi ID
     */
    void deleteWaitingList(Long id);
    
    /**
     * Kullanıcıyı bekleme listesinden çıkarır
     * 
     * @param eventId Etkinlik ID
     * @param userId Kullanıcı ID
     */
    void removeFromWaitingList(Long eventId, Long userId);
    
    /**
     * Etkinlik için bekleme listesindeki bir kişiyi katılımcı olarak ilerletir
     * 
     * @param waitingListId Bekleme listesi ID
     * @return Güncellenen bekleme listesi kaydı
     */
    WaitingListDto promoteToParticipant(Long waitingListId);
    
    /**
     * Etkinlik için bekleme listesindeki bir kişiyi katılımcı olarak ilerletir
     * WaitingListController için kullanılacak
     * 
     * @param eventId Etkinlik ID
     * @param userId Kullanıcı ID
     */
    void promoteToParticipant(Long eventId, Long userId);
    
    /**
     * Bekleme listesinin sırasını değiştirir (öncelik ayarı)
     * 
     * @param eventId Etkinlik ID
     * @param waitingListItems Yeni sıralama ile bekleme listesi kayıtları
     * @return Güncellenen bekleme listesi kayıtları
     */
    List<WaitingListDto> reorderWaitingList(Long eventId, List<WaitingListDto> waitingListItems);
    
    /**
     * Etkinlik için kontenjan açıldığında bekleme listesindeki kişilere otomatik bildirim gönderir
     * 
     * @param eventId Etkinlik ID
     * @param availableSlots Açılan kontenjan sayısı
     * @param responseDeadlineHours Yanıt süresi (saat)
     * @return Bildirim gönderilen bekleme listesi kayıtları
     */
    List<WaitingListDto> notifyWaitlistForAvailableSlots(Long eventId, int availableSlots, int responseDeadlineHours);
    
    /**
     * Bekleme listesindeki kişileri otomatik olarak katılımcı listesine ekler
     * 
     * @param eventId Etkinlik ID
     * @param limit Maksimum eklenecek kişi sayısı (0=sınırsız)
     * @return Katılımcı olarak eklenen kişi sayısı
     */
    int autoPromoteFromWaitingList(Long eventId, int limit);
} 